gdal/ogr/ogrsf_frmts/cad/ogrcaddriver.cpp

```diff
@@ -0,0 +1,148 @@
+    {
+        char** papszTokens = CSLTokenizeString2(poOpenInfo->pszFilename, ":", 0);
+        if( CSLCount(papszTokens) != 4 )
+        {
+            CSLDestroy(papszTokens);
+            return FALSE;
+        }
+
+        pFileIO = new VSILFileIO( papszTokens[1] ); 
+        nSubRasterLayer = atol(papszTokens[2]);
+        nSubRasterFID = atol(papszTokens[3]);
+        
+        CSLDestroy(papszTokens);
+    }
+    else
+    {
+        pFileIO = new VSILFileIO( poOpenInfo->pszFilename );    
+    }
+    
+    if ( IdentifyCADFile( pFileIO, false ) == FALSE)
+    {
+        delete pFileIO;
+        return NULL;
+    }
+        
+        
+/* -------------------------------------------------------------------- */
+/*      Confirm the requested access is supported.                      */
+/* -------------------------------------------------------------------- */
+    if( poOpenInfo->eAccess == GA_Update )
+    {
+        CPLError( CE_Failure, CPLE_NotSupported,
+                  "The CAD driver does not support update access to existing"
+                  " datasets.\n" );
+        delete pFileIO;
+        return NULL;
+    }
+         
+    GDALCADDataset *poDS = new GDALCADDataset();    
+    if( !poDS->Open( poOpenInfo, pFileIO, nSubRasterLayer, nSubRasterFID ) )
+    {
+        delete poDS;
+        return NULL;
+    }
+    else
+        return poDS;
+}
+
+/************************************************************************/
+/*                           RegisterGDALCAD()                          */
+/************************************************************************/
+
+void RegisterOGRCAD()
+{
+    GDALDriver  *poDriver;
+    
+    if ( GDALGetDriverByName( "CAD" ) == NULL )
+    {
+        poDriver = new GDALDriver();
+        poDriver->SetDescription( "CAD" );
+        poDriver->SetMetadataItem( GDAL_DCAP_RASTER, "YES" );
+        poDriver->SetMetadataItem( GDAL_DCAP_VECTOR, "YES" );
+        poDriver->SetMetadataItem( GDAL_DMD_SUBDATASETS, "YES" );
+        poDriver->SetMetadataItem( GDAL_DMD_LONGNAME, "AutoCAD Driver" );
+        poDriver->SetMetadataItem( GDAL_DMD_EXTENSION, "dwg" );
+        poDriver->SetMetadataItem( GDAL_DMD_HELPTOPIC, "drv_cad.html" );        
+        
+        poDriver->SetMetadataItem( GDAL_DMD_OPENOPTIONLIST, "<OpenOptionList>"
+"  <Option name='MODE' type='int' min='1' max='3' description='Open mode. 1 - read all data (slow), 2 - read main data (fast), 3 - read less data' default='2'/>"
+"</OpenOptionList>"); 
+
+        
+        poDriver->pfnOpen = OGRCADDriverOpen;
+        poDriver->pfnIdentify = OGRCADDriverIdentify;
+        poDriver->SetMetadataItem( GDAL_DCAP_VIRTUALIO, "YES" );
+        GetGDALDriverManager()->RegisterDriver( poDriver );
+    }
+}
+
+CPLString CADRecode( const std::string& sString, int CADEncoding )
+{
+    switch( CADEncoding )
+    {
+        case 29:
+            return CPLString( CPLRecode( sString.c_str(), "CP1251", CPL_ENC_UTF8 ) );
+    }
```

Comment:
Temporarily fixed. I could not find all CADEncoding values mapping into encoding name, so by default it will be giving a warning about unsupported encoding for now.